.github/workflows/tests.yml

```diff
@@ -2,22 +2,117 @@ name: Tests
 
 on:
   push:
-    branches: [master]
+    branches: ['*']
   pull_request:
     branches: [master]
 
 jobs:
+  cleanup:
+    name: Clean test images
+    needs:
+    - test-ui
+    - test-api
+    runs-on: ubuntu-latest
+    steps:
+    - name: Cleanup
+      uses: actions/delete-package-versions@v5
+      with:
+        owner: '${{ github.repository_owner }}'
+        token: '${{ secrets.GITHUB_TOKEN }}'
+        package-type: 'container'
+        package-name: photoview/test
+        min-versions-to-keep: 5
```

Comment:
I tried with `actions/delete-package-versions`, but the `package-version-ids` parameter requires an integer id instead of the tag. See https://github.com/actions/delete-package-versions/issues/101. I also tried to query GitHub API to get such id. The problem is [GitHub package API](https://docs.github.com/en/rest/packages?apiVersion=2022-11-28) requires an organization or a user to query packages, not a repo. To check a repo belonging to an organization or a user, the workflow is far more complex and I'm not happy about such complexity. I know the problem you said.
I don't know if it works with Docker Hub. With Docker Hub, all forked repos can't run the test workflow themselves. I hope not to do that.

Reply:
I tried some more different methods and realized that all actions are wrapped scripts running in the runner under `ubuntu-latest`. We could combine those actions and run scripts after that. Finally, I got [this workflow](I tried a few different methods). It cached all layers when building test images and code dependencies, then ran tests in containers. It doesn't push images to any repos so no need to clean up. It still runs tests with `docker compose` commands, but not through a declared container. As GitHub runner declares [providing Docker tools](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#tools) with `ubuntu-latest`, I read it as GitHub supports such feature officially.
Let me know what you think about it.

Reply:
I got [this workflow] - did you forget to put a link to some other place here?

Reply:
> To check a repo belonging to an organization or a user, the workflow is far more complex

Yeap, I spent days to make it work in my forked repo, and I think that in the case of merging that PR into the parent repo, I'd have to spend even more time to make it work for it(

I need to see what is the alternative workflow, you mentioned in the 2nd comment, but if we go with this one, at least increase # of versions left to 20 so that we support up to 10 running in-parallel jobs (I guess that this would be more than enough taking into account Github CI performance and how often PRs are changed). Also, please add the step to remove all untagged versions - any job does not use them and they just use the space.

Reply:
> > I got [this workflow]
> 
> Did you forget to put a link to some other place here?

Sorry, copy/paste the wrong content. See [here](https://github.com/googollee/photoview/actions/runs/10151697901/workflow).

Reply:
I prefer [that compose workflow](https://github.com/googollee/photoview/actions/runs/10157776885/workflow). It's faster than current one. If missing cache, it takes 5m for API test and 2m for UI test. If hitting cache, it takes 2m for API test and 1m for UI test. As the compose workflow doesn't share images between jobs, it doesn't need to upload/download/cleanup images, which saving a lot of time and reducing the complexity.

Reply:
yes, but it introduces more complexity in the test execution setup - the main part of the testing job)
from my point of view, we need to provide the clearest possible setup, as in an open-source project it should be clear to even a junior-level developer (with the help from Google and GPT) how the workflow works and to debug in the case of issues.
Even if the local test execution setup is not so clear for such developers, the CI should be the right place to understand the flow and debug it.

I already explained this earlier, so won't add much detail now.

Reply:
How about expand the command in workflow? Like below:

```
    - name: Test
      run: docker compose -f dev-compose.yaml run test-api-${{ matrix.database }}

    - name: Cleanup
      run: docker compose -f dev-compose.yaml down -v
```

It shows more details in the CI workflow.

Reply:
not a big difference, as to me...

Reply:
Outdated.